clang/lib/Sema/SemaChecking.cpp

```diff
@@ -5553,6 +5557,58 @@ bool Sema::BuiltinSetjmp(CallExpr *TheCall) {
   return false;
 }
 
+bool Sema::BuiltinCountedByRef(CallExpr *TheCall) {
+  // For simplicity, we support only a limited expressions for the argument.
+  // Specifically 'ptr->array' and '&ptr->array[0]'. This allows us to reject
+  // arguments with complex casting, which really shouldn't be a huge problem.
+  if (checkArgCount(TheCall, 1))
+    return true;
+
+  ExprResult ArgRes = UsualUnaryConversions(TheCall->getArg(0));
+  if (ArgRes.isInvalid())
+    return true;
+
+  const Expr *Arg = ArgRes.get()->IgnoreParenImpCasts();
+  if (!isa<PointerType>(Arg->getType()) && !Arg->getType()->isArrayType())
+    return Diag(Arg->getBeginLoc(),
+                diag::err_builtin_counted_by_ref_must_be_flex_array_member)
+           << Arg->getSourceRange();
+
+  if (Arg->HasSideEffects(Context))
+    Diag(Arg->getBeginLoc(), diag::warn_builtin_counted_by_ref_has_side_effects)
+        << Arg->getSourceRange();
+
+  // See if we have something like '&ptr->fam[0]`.
```

Comment:
Should we be kind for anyone doing `&(ptr->fam[0])` as well?

Reply:
I want to strangle anyone who does that, but I suppose you have a point. I hate that the AST we're working on during CodeGen still has such cruft in it instead of preparing the AST ready for real code generation.

Reply:
Also, the `IgnoreParenImpCasts` should ignore parens. At least that's what they say it does.

Reply:
Oh you're right, I missed we were already ignoring parens. Carry on.